libs/ballot-interpreter-nh-next/src/timing_marks.rs

```diff
@@ -0,0 +1,1038 @@
+///   1. rotate 180 degrees if `orientation` is `PortraitReversed`.
+///   2. crop the image edges by `border_inset`.
+///   3. scale the image to `scaled_size`.
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TimingMarkGrid {
+    /// The geometry of the ballot card.
+    pub geometry: Geometry,
+
+    /// The orientation of the ballot card.
+    pub orientation: Orientation,
+
+    /// Inset to crop to exclude the border.
+    pub border_inset: Inset,
+
+    /// The size of the image after scaling.
+    pub scaled_size: Size<u32>,
+
+    /// Timing marks found by examining the image.
+    pub partial_timing_marks: Partial,
+
+    /// Timing marks inferred from the partial timing marks.
+    pub complete_timing_marks: Complete,
+
+    /// Areas of the ballot card that contain shapes that may be timing marks.
+    pub candidate_timing_marks: Vec<Rect>,
+
+    /// Metadata from the ballot card bottom timing marks.
+    pub metadata: BallotPageMetadata,
+}
+
+impl TimingMarkGrid {
+    pub fn new(
+        geometry: Geometry,
+        orientation: Orientation,
+        border_inset: Inset,
+        scaled_size: Size<u32>,
+        partial_timing_marks: Partial,
+        complete_timing_marks: Complete,
+        candidate_timing_marks: Vec<Rect>,
+        metadata: BallotPageMetadata,
+    ) -> Self {
+        Self {
+            geometry,
+            orientation,
+            border_inset,
+            scaled_size,
+            partial_timing_marks,
+            complete_timing_marks,
+            candidate_timing_marks,
+            metadata,
+        }
+    }
+
+    /// Returns the center of the grid position at the given coordinates. Timing
+    /// marks are at the edges of the grid, and the inside of the grid is where
+    /// the ovals are.
+    ///
+    /// For example, if the grid is 34x51, then:
+    ///
+    ///   - (0, 0) is the top left corner of the grid
+    ///   - (33, 0) is the top right corner of the grid
+    ///   - (0, 50) is the bottom left corner of the grid
+    ///   - (33, 50) is the bottom right corner of the grid
+    ///   - (c, r) where 0 < c < 33 and 0 < r < 50 is the oval at column c and
+    ///     row r
+    pub fn point_for_location(&self, column: u32, row: u32) -> Option<Point<f32>> {
+        if column >= self.geometry.grid_size.width || row >= self.geometry.grid_size.height {
+            return None;
+        }
+
+        let timing_mark_width = self.geometry.timing_mark_size.width.round() as u32;
+        let left = self.complete_timing_marks.left_rects.get(row as usize)?;
+        let right = self.complete_timing_marks.right_rects.get(row as usize)?;
+        let corrected_left = Rect::new(
+            left.right() - timing_mark_width as i32,
+            left.top(),
+            timing_mark_width,
+            left.height(),
+        );
+        let corrected_right =
+            Rect::new(right.left(), right.top(), timing_mark_width, right.height());
```

Comment:
Maybe add a comment explaining why correction is necessary?

Reply:
Added.